### ISSUE TITLE
[feature request(collections)] option to merge `undefined` in `deepMerge`

### ISSUE THREAD

**Issue by scarf005** [denoland/deno_std#4752]
**Is your feature request related to a problem? Please describe.**

when merging together deeply nested object (e.g react state), it's important to get 'nonempty' values when merging 'empty' values with 'nonempty' values.

for array, map and set it defaults to merging empty and non-empty values together:

```ts
deepMerge({ a: new Set([1]) }, { a: new Set() })
{ a: Set(1) { 1 } }
```

however, it doesn't work with nullish values. it'd be very useful to be able to output non-nullish value from merging nullish and non-nullish value together.

```ts
deepMerge({ a: 1 }, { a: undefined })
// expected: { a: 1 }
// actual: { a: undefined }
```

**Describe the solution you'd like**

in [DeepMergeOptions](https://jsr.io/@std/collections@0.224.1/doc/deep-merge/~/DeepMergeOptions), add `nullish: MergingStrategy` that controls whether `deepMerge` will prefer non-nullish value over nullish ones.

**Describe alternatives you've considered**

make `nullish: "merge”` as default. this may cause breaking changes.

**Comment by kt3k** [comment]
Does `nullish: "merge"` mean that `null` also has the same effect as `undefined` in the above context?

**Comment by scarf005** [comment]
> Does `nullish: "merge"` mean that `null` also has the same effect as `undefined` in the above context?

yes, it could have been `nullish: "replace" | "mergeBoth" | "mergeNull" | "mergeUndefined"` but the type would be too complex.

**Comment by kt3k** [comment]
How about only doing this for `undefined`? (The option would be `undefined: "replace" | "ignore"`

Handling null in the same way feel a bit strange to me as `{ a: null }` is not assignable to `Partial<{ a: number }>` for example.

**Comment by scarf005** [comment]
sure, since `Partial` only works with `undefined`.
